Add 2022 review post (#397)
trinodb/trino.io: _posts/2023-01-04-trino-2022-the-rabbit-reflects.md

you still have to do, and that isn’t healthy for anybody. In that spirit, let’s
reflect on what we’ve been able to accomplish as a community this year, as well
as what to look forward to in the next year!

<!--more-->

## 2022 by the numbers

Let’s take a look at the Trino project’s growth and what happened specifically
in the past year:

* 1,031,842 unique visits 🙋to the Trino site
* 116,231 unique blog post views 👩‍💻 on the Trino site
* 60,296 views 👀 on YouTube
* 5,982 hours watched ⌚on YouTube
* 4,696 new commits 💻 in GitHub
* 2,775 new members 👋 in Slack
* 2,769 new stargazers ⭐ in GitHub
* 2,550 pull requests merged ✅ in GitHub
* 1,465 issues 📝 created in GitHub
* 1,322 new followers 🐦 on Twitter
* 1,068 pull requests closed ❌ in GitHub
* 702 new subscribers 📺 in YouTube
* 658 average weekly members 💬 in Slack
* 56 videos 🎥 uploaded to YouTube
* 37 Trino 🚀 releases
* 36 blog ✍️ posts
* 12 Trino Community Broadcast ▶️ episodes
* 12 Trino 🍕 meetups
* 2 Trino ⛰️ Summits

The Trino website got an impressive number of unique visits, also referred to as
entrances. This metric filters out refreshes and through traffic to count the

> number of times a visitor started a unique session. Blog posts saw a 47 percent

Author: Google style says to spell it out when using it in a sentence.

 `LIKE` expressions and highly selective `LIMIT` queries, and enhanced
 performance and reliability for `INSERT` and `MERGE` operations. We also made
 performance improvements for `JOIN`, `UNION`, and `GROUP BY` queries, as well
 as faster planning of queries with `IN` predicates. 
* There are also optimizations for specific SQL types’ performance, such as
 string, `DECIMAL`, `MAP` and `ROW` types. We have also made aggregations over 
 `DECIMAL` columns faster and improved the performance of `ROW` type and
 aggregation.
* A last set of improvements come from reading open file formats like ORC and
 Parquet efficiently. We have improved the speed of reading or writing of all 
 data types from and to Parquet in general. There were also general performance
 to ORC types, and now have the ability to write Bloom filters in ORC files. We
 have also improved performance and efficiency for a wide range of ORC and
 Parquet-related operations.

These improvements in aggregate are at the core of what makes Trino fast. There
is no silver bullet you can plug in to speed things up. It takes time, effort,
and smart changes to improve the speed of various systems.

### Runtime improvements

> [Trino upgraded Java versions](https://github.com/trinodb/trino/issues/9876)from

Reviewer: ```suggestion
[Trino upgraded the required Java version](https://github.com/trinodb/trino/issues/9876) from
```

> Java 11 to the next long-term support version of Java, which was Java 17. This

Reviewer: ```suggestion
Java 11 to the current long-term support - Java 17. This
```

> upgrade improves the latency, memory footprint, startup, ramp up, pause times,

Reviewer: kinda odd .. most of them are "reduced" .. not" improved"

> and shut down times for Trino. Along with these performance enhancements one of the other

Reviewer: The rest is too long .. essentially it just needs to say that the GC in Java 17 is also improved

critical elements of improving performance for the Java Virtual Machine is to
improve the garbage collector. Trino uses the G1 garbage collector which can now
more efficiently reclaim memory and reduce pause times.

> Martin did a benchmark on Trino workloads, and reported seeing anywhere from

Reviewer: I would drop nearly all of this .. we just need to say that the upgrade has not caused any problems and went smoothly and we got a bunch of performance improvements for free

10-15% improvement in speed in Java 17. Many users have also reported a similar
boost in their workloads, and we have yet to have a user mention any regressions
related to Java 17, which is exciting news. Aside from having to perform the
upgrades, we get a lot of these performance enhancements for free.

> On top of performance, upgrading to Java 17 adds new Java language features to

Reviewer: Cut this down to just one sentence as well about language and API improvements now being available for use.. and forget about the vector stuff and more details

improve the ability to write and maintain higher quality code. Which matters a
lot in a fast-moving open-source project like Trino. Java 17 also adds a Vector
API, which allows Trino to explore running some algorithms on native hardware,
ideally resulting in a speed-up from Java's exceptional auto-vectorization.

To learn more, read [this blog 

> post]({% post_url 2022-07-14-trino-updates-to-java-17 %}) and watch this episode

Reviewer: ```suggestion
post]({% post_url 2022-07-14-trino-updates-to-java-17 %}) and watch episode 36
```

and make the link longer or just "episode 36"

of [the Trino Community Broadcast]({{site.url}}/episodes/36.html)

Along with the Java upgrade, Trino now has a Docker image for ppc64le and added
CLI support for ARM64, which means Trino's Docker image and the CLI can run on

> the new MacBooks.

Reviewer: and also more importantly on ARM chip cloud stuff

  
### Security

Trino added the following improvements and features relevant for authentication,
authorization and integration with other security systems:

* There were a lot of updates to [OAuth 2.0
 authentication]({{site.url}}/docs/current/security/oauth2.html) like support for OAuth
 2.0 refresh tokens and allowing access token passthrough with refresh tokens
 enabled. We also added support for [automatic discovery of OpenID
 Connect]({{site.url}}/docs/current/security/oauth2.html#openid-connect-discovery)

>  metadata with OAuth 2.0 authentication. Support for groups in OAuth2 claims and

Reviewer: ```suggestion
 metadata with OAuth 2.0 authentication, Support for groups in OAuth2 claims, and
```

 reduced latency for OAuth2.0 authentication.

> * Access control rules got a fair amount of attention. Support for string

Reviewer: I feel like this whole item could be dropped

Author: dropped

---
layout: post

> title: "Trino 2022 the Trino rabbit reflects"

Author: How about Trino 2022: The rabbit reflects

> title: "Trino 2022 the Trino rabbit reflects"

Author: The rabbit reflects on Trino in 2022
I like this one

hope sharing this impact motivates you to help us. We don’t want to keep Trino
the hidden secret of Silicon Valley much longer. We need your help to really get
people continuously reading and hearing about all things Trino. Even if
individual shares don’t get much engagement, remember views on your posts count,
too, as many silently click on links. So share any time you see something cool
going on in our community!

### Trino touches the world

Let’s take a look at the number of users who have initiated at least one session
on the Trino site in 2022 by top 10 countries. This goes to show the true global
reach this project has attained in 10 years.

* 123,326 USA 🇺🇸users
* 33,540 Indian 🇮🇳users
* 30,955 Chinese 🇨🇳users
* 12,282 British 🇬🇧users
* 11,638 German 🇩🇪users
* 10,760 Canadian 🇨🇦 users
* 9,980 Brazilian 🇧🇷users
* 9,098 Singaporean 🇸🇬users
* 8,649 South Korean 🇰🇷users
* 8,636 Japanese 🇯🇵users

![](/assets/blog/2022-review/world.png)

Our reach currently favors the USA, but our aim is to grow Trino in all
countries that are starting to show a stronger interest in it. The new edition
of “Trino: The Definitive Guide” is being translated into Chinese, [Polish, and

> Japanese.](https://simpligility.ca/2022/12/trino-guide-for-everyone-in-2023/) If

Reviewer: end of sentence . should not be part of the link

Manfred Moser.

## Trino celebrates its tenth birthday

Of all the incredible things that happened this year, one that gave us pause
earlier this year was Trino’s tenth birthday. Martin, Dain, and David [cite
longevity]({{site.url}}/development/vision.html) of the project as one of
the core philosophies that govern decisions around Trino. We expect that Trino
will be used for at least the next 20 years. We build for the long term. This
first decade [has been an adventurous ride
]({% post_url 2020-12-27-announcing-trino %}), and wow has it
[produced an incredible system]({% post_url 2022-08-08-trino-tenth-birthday %}).

![](/assets/blog/trino-tenth-birthday/how-it-started-going.png)

We wanted to do something special with the community to celebrate this
milestone, so Brian put together a birthday video to timeline the evolution of
Presto and now Trino. We had a premiere watch party on the day of the tenth
anniversary and got some folks' reactions. Take a look at the video if you
haven’t yet, you don’t want to miss it.

{% youtube hPD95_-bZZw %}

## Trino Summit

The next event in 2022 was the Trino Summit, which was the first in-person

> summit we’ve had, with well over 500 attendees. We had a stellar lineup of

Author: I meant to say "as Trino"

amount of interaction from those new voices.

![](/assets/blog/2022-review/virtual-tcc.png)

During these discussions the biggest topics covered timelines of existing
roadmap items and suggestions for other items that should get more attention.
We covered upcoming connectors and plugins, and all the required infrastructure
needed to support that. A recurring theme was the need for better testing
infrastructure. The more information we can gather as a community, the quicker
we can remove any issues as new releases come out and increase adoption of
newer versions of Trino. We also discussed desired features around
resource-intensive and batch workloads, and the new polymorphic table function
features.

The biggest takeaway from these meetings was that everyone now had a better
basis to engage with each other. As we move forward, we will continue the
cadence of having these virtual TCCs to keep everyone on the same page, and have
in-person meetings when there is a larger conference. With that, let’s cover
some of the features we gained this year.

## Features

Of course, one of the main deliverables of our project are Trino releases. In

> 2022 we improved our release process and cadence shipped 37 releases that were

Reviewer: ```suggestion
2022 we improved our release process and cadence, and shipped 37 releases that were
```

packed with features. Here’s a high-level list of the most exciting ones that
made their way to you. For details and to keep up you can check out the
[release notes]({{site.url}}/docs/current/release.html).

### Fault-tolerant execution mode

2022 was the year of resiliency for Trino. One of the longest requested features

> that our user had was adding a [fault-tolerant mechanism to Trino

Reviewer: ```suggestion
that our users had was adding a [fault-tolerant mechanism to Trino
```

]({{site.url}}/docs/current/admin/fault-tolerant-execution.html) akin to
query engines like Apache Spark. Users wanted the ability to take the queries
that they were running in Trino, and scale those queries to larger data or
resource intensive queries. Experimental features were implemented in late 2021
for [automatic query retries](https://github.com/trinodb/trino/pull/9361) and
earlier this year [task-level retries
](https://github.com/trinodb/trino/pull/9818). The efforts for these features
were codenamed [Project Tardigrade]({{site.url}}/episodes/32.html).

This feature relies on storing intermediate data between task shuffles to
persist data in what was called an exchange spool. The first iteration of this
was AWS S3, but eventually Azure Blob Storage, Google Cloud Storage, and local

> storage were included. As the Project Tardigrade engineers started [improving

Reviewer: lets leave local storage out ..

Reviewer: also ... this is not a sentence at this stage..

performance and fixing bugs]({% post_url 2022-02-16-tardigrade-project-update %})
in fault-tolerant execution as they and a few users tested the early
implementation. Later, memory efficiency for aggregations, faster data
transfers, and dynamic filtering with fault-tolerant query execution were added.
The [launch of fault-tolerant execution
]({% post_url 2022-05-05-tardigrade-launch %}) launch happened at
Cinco de Trino. The first iterations only applied to queries being run on
object-storage connectors such as Hive, Iceberg, and Delta Lake. Recently,
support for MySQL, PostgreSQL, and SQL Server were added. These contributions

> added a foundation for other JDBC connectors.

Reviewer: need to add a final sentence about adoption

### SQL language improvements

> Here are all the notable SQL features that made it in to Trino this year:

Reviewer: ```suggestion
Here are all the notable SQL features that made it to Trino this year:
```

* [`MERGE` statement support]({{site.url}}/docs/current/sql/merge.html) is
 the most impactful SQL feature released this year. `MERGE` allows users to

>  implement `INSERT`, `UPDATE`, and `DELETE` statements in one fell swoop.

Reviewer: ```suggestion
 implement `INSERT`, `UPDATE`, and `DELETE` functionality in one statement.
```

>  `MERGE` is not simply syntax sugar, it actually has profound performance

Reviewer: ```suggestion
 `MERGE` is not simply syntax sugar, the implementation includes profound performance
```

>  implications as now a lot of your operations can be merged (pun intended) from

Reviewer: ```suggestion
 impprovements. Now a lot of your operations can be merged (pun intended) from
```

 multiple tasks into a single scan over data. This functionality is absolutely
 critical for positioning Trino as a data lakehouse query engine. `MERGE` is 
 currently available in Hive, Iceberg, Delta Lake, Kudu, and Raptor connectors.
 We discussed this and did a demo with `MERGE` on the recent [Trino Community
 Broadcast with Iceberg]({{site.url}}/episodes/40.html).
* Another massive update was the introduction of [polymorphic table functions
 ]({% post_url 2022-07-22-polymorphic-table-functions %}) ([PTFs
 ]({{site.url}}/docs/current/functions/table.html)). Table functions
 initially released with some initial passthrough query functionality that we

>  see in connectors like Pinot, Elasticsearch, MySQL, PostgreSQL, and many

Reviewer: ```suggestion
 see in connectors like Pinot, Elasticsearch, MySQL, PostgreSQL, and
```

Author: what about "and other JDBC connectors" and I'll link to the PR that points to all of them

 others. However, this is only one small instance of what can be achieved with
 PTFs and the [true power comes from the generalization of this feature
 ](https://www.youtube.com/clip/UgkxQcokpdgPjiuMKMC5-3HwHvlbmZjxAvxe). Dain and
 David gave [a simpler explanation of PTFs
 ](https://www.youtube.com/clip/Ugkx62IKgPd_v9eGBaPUHP2hyaRkWSXh8w8h) if it’s
 still not very clear what they are. To dive in deeper, watch [this episode of
 the Trino Community Broadcast]({{site.url}}/episodes/38.html) where Kasia
 Findeisen and Martin discuss PTFs in greater detail.
* [Dynamic function resolution](https://github.com/trinodb/trino/issues/8) has

>  been discussed for many years. This provides the ability for [connectors to

Reviewer: ```suggestion
 been discussed for many years and finally arrived. This provides the ability for [connectors to
```

 provide functions at runtime](https://youtu.be/mUq_h3oArp4?t=680). Unlike
 before, where you needed to statically register your functions ahead of time,

>  you can now provide a plugin that defines these functions that gets resolves

Reviewer: ```suggestion
 you can now provide a plugin that provides functions hat are resolved at runtime.
```

Author: I don't like the "provide" sandwhich. Maybe provide a plugin that contains...

 while Trino is running. This enables features like supporting function calls to
 dynamically registered user-defined functions in different languages like
 Javascript or Python. Martin and Dain go into great detail about how this works
 when [answering this question at Trino Summit
 ](https://youtu.be/mUq_h3oArp4?t=1596).

> * [ANSI SQL 2016](https://en.wikipedia.org/wiki/SQL:2016) support for JSON 

Reviewer: Added support for some JSON processing function, which are part of the ANSI 2016 specification.

 functions resolves a large number of issues reported by the community over the
 years. This includes 
 [json_array]({{site.url}}/docs/current/functions/json.html#json-array),
 [json_object]({{site.url}}/docs/current/functions/json.html#json-object),
 [json_exists]({{site.url}}/docs/current/functions/json.html#json-exists),
 [json_query]({{site.url}}/docs/current/functions/json.html#json-query), and
 [json_value]({{site.url}}/docs/current/functions/json.html#json-value)
 functions that were added to Trino this year.
* The JSON output format was added to the `EXPLAIN` statement to provide

>  anonymised query plan to enable offline analysis.

Reviewer: Change to mention both JSON format and anoymized output .. but as separate things

Author: just gonna say JSON format vs JSON format Output

* It became possible to comment on tables, columns of tables, and even views for
 various connectors. Support for setting comments on views was just introduced

>  recently and includes support for Hive and Iceberg. Table comments came to

Reviewer: We can drop the list of connectors imho

Author: fair...comments aren't mind blowing features..just nice

 Delta Lake, MySQL, BigQuery, Clickhouse and column comments came to Iceberg,
 Accumulo, MongoDB, BigQuery, Clickhouse.
* There were also a lot of new functions added like `to_base32`, `from_base32`,
 `trim_array`, `trim`, and many others.

### Performance improvements

Despite all the hype about vectorization being a silver bullet to make databases
go fast, the real speed comes from [better algorithms and better data structures
that lead to lower resource consumption](https://www.youtube.com/clip/UgkxQwDYDS6evVJelNVjWAgrIhzg_Q-cAEyq).

> Here are a few that made their way into Trino this year:

Reviewer: ```suggestion
Following is a list of some of the improvements that made their way into Trino:
```

* Trino now offers improved performance for a variety of operations, including
 complex join criteria pushdown to connectors, faster aggregations, faster
 joins, and better performance for large clusters. We have also implemented
 improvements specifically for aggregations with filters and for the Glue
 metastore. In addition, we now support dynamic filtering for various connectors
 and have faster query planning for Hive, Delta Lake, Iceberg, MySQL,
 PostgreSQL, and SQL Server connectors.
* Along with general performance optimizations, there have been a great deal of
 query planning optimizations that lead to better performance for specific SQL
 operators. These include faster `INSERT` queries, improved performance for
 `LIKE` expressions and highly selective `LIMIT` queries, and enhanced
 performance and reliability for `INSERT` and `MERGE` operations. We also made
 performance improvements for `JOIN`, `UNION`, and `GROUP BY` queries, as well

>  as faster planning of queries with `IN` predicates. We have also added support

Reviewer: Drop last sentence

 for correlated subqueries in `DELETE` queries.

> * There are also optimizations for specific SQL types’ performance like of

Reviewer: There are also optimizations for specific SQL types’ performance such as string, decimal, MAP, and ROW types.

 string processing, `MAP` and `ROW` types, the `DECIMAL` type, and `ROW` data
 types. We have also made aggregations over `DECIMAL` columns faster and
 improved the performance of `ROW` type and aggregation.

> * A last set of improvements come from reading open file formats like ORC and

Reviewer: This whole paragraph should be shortened to something to get the idea across that both ORC and Parquet processing is now very fast. Specifically Parquet had significant improvements.

 Parquet efficiently. We have improved the speed of reading decimal values in
 Parquet data, writing `ARRAY`, `MAP`, and `ROW` types to Parquet, and reading
 Parquet data in general. We have also made writing non-structural types to
 Parquet faster, providing improved performance for a wide range of
 Parquet-related operations. ORC data more efficiently. We have made
 improvements to the performance of ORC in general, and have added the ability
 to write Bloom filters in ORC files, and implemented these filters for
 `VARCHAR` columns, further improving the efficiency of ORC-related operations.

These improvements in aggregate are at the core of what makes Trino fast. There
are rarely silver bullets like vectorization that can improve the overall speed.

### Runtime improvements

The biggest overall runtime improvement comes from the Java ecosystem. Java 17 was the next LTS from Trino’s existing version Java 11.

* Java 17 (390) {{site.url}}/episodes/36.html {% post_url 2022-07-14-trino-updates-to-java-17 %} https://github.com/trinodb/trino/issues/9876
* Experimental docker image for ppc64le. (393)

> * Add CLI support for ARM64. (370)

Author: Still working on the runtime improvements section

> ### Security

Reviewer: Seems like a very long section now .. can we shorten?

Author: Removed the last bullet

 effective lakehouse query engine. Iceberg also now supports the Glue.
    * Faster `date_trunc` predicates over partition columns in Iceberg connector. (397)
    * Reduced query latency with Glue catalog in Iceberg connector. (397)
    * Support for Avro format in Iceberg connector. (392)
    * Support for time travel syntax in the Iceberg connector.(385)
    * Performance improvements in Iceberg connector.(385)
    * Improved latency when querying Iceberg tables with many files. (401)
    * Support for `UPDATE` in Iceberg connector. (381)
    * Allow changing file and table format with the Iceberg connector. (380)
    * File-based access control for the Iceberg connector. (367/368)
* Although it is seeming more and more like Hive is already on its way out,
 there are many that still depend on the Hive connector to be performant and
 adding new features when it makes sense. S3 Select pushdown for JSON data in Hive connector.
    * Support for IBM Cloud Object Storage in Hive. (401)
    * Rename properties controlling Hive view parsing. (380)
    * Improved performance when querying partitioned Hive tables. (373)

>     * `flush_metadata_cache()` procedure for the Hive connector. (367/368)

Author: Still working on the object storage connectors section

* Performance improvements with various pushdowns in the MongoDB, MySQL, Oracle,
 PostgreSQL and SQL Server connectors.
* Support for bulk data insertion in SQL Server connector.
* Added a query passthrough table function to numerous connectors.
* Expanded SQL features for various connectors by adding support for
 `TRUNCATE TABLE`, `DELETE`, `CREATE`/`DROP` `SCHEMA`, `INSERT`, and others.
* Update Cassandra connector to support v5 and v6 protocols.
* A collection of improvements on the Pinot and BigQuery connectors

### Bug fixes

Any software includes issues and bugs, Trino included. Thanks to our big user
community we learned about many of them, and fixed even more. Continue to test
new releases and report issues. Check out[ all the release notes for details
]({{site.url}}/docs/current/release.html#releases-2022).

## Updates in the Trino ecosystem community

Outside of the excitement within the main Trino project, there was a great deal
going on in the larger Trino community and ecosystem:

### Trino: The Definitive Guide second edition

Martin, Manfred, and Matt released the [second version of the Trino Definitive
Guide]({% post_url 2022-10-03-the-definitive-guide-2 %}). This 

> version fixed errata, added the deployment process to include newer Kubernetes

Reviewer: ```suggestion
update of the book from O'Reilly fixed errata, added the deployment process to include newer Kubernetes
```

installation methods, and updated features with the latest version of Trino that
was not available in the initial version of the book. Along with this, there are

> [efforts underway to translate this book](https://simpligility.ca/2022/12/trino-guide-for-everyone-in-2023/)

Reviewer: we either remove it here .. or in the section about regions earlier

Author: Why though? It's such a long blog post that they may find skip one section and only see the other. I think it's relevant to both.

Author: I can remove the link on the one above, but I think that we should mention it in both places.

to different languages. Huge thanks to everyone involved in this effort!

### Starburst provides Trino in the cloud

As a major community supporter, [Starburst](https://starburst.io/) helped us
with events, marketing, developer relations, and partner cooperation. Starburst
also provided a large part of development and code contributions to Trino and
its related projects. Starburst acquired Varada and integrated the object
storage indexing technology, and they shipped many Starburst Enterprise releases
for self-managed deployments. On top of all that amazing work, Starburst
launched [Starburst Galaxy](https://www.starburst.io/platform/starburst-galaxy/)

> as a powerful, muli-cloud SaaS offering of Trino. Security, cluster management,

Reviewer: ```suggestion
as a powerful, multi-cloud SaaS offering of Trino. Security, cluster management,
```

a query editor, and many other features are included in this new platform.

### Amazon upgrades Athena

[Athena version three rolled out]({% post_url 2022-12-01-athena %})
and is now based on a recent Trino release. This is great news for Athena users
who were missing the many performance gains, expanded SQL support, and other
features from Trino, since the prior versions are based on old Presto releases.
As a result, the large Athena community and their feedback and knowledge have
become more integrated with the Trino community, and we are seeing positive
impact for Trino releases already.

### dbt-trino

dbt users rejoice! The [official dbt Trino integration
](https://docs.getdbt.com/reference/warehouse-setups/trino-setup) made it into
dbt this year! This means that anyone who wanted to read or write data from or
to multiple data sources is now able to. If you want to dive into all you need
to know to get started and a tutorial, [check out this blog post
](https://docs.starburst.io/blog/2022-11-30-dbt0-introduction.html) written by
the contributors of this integration.

### Python client improvements

> The trajectory of the [trino-python-client

Reviewer: wrapping here might break link

](https://github.com/trinodb/trino-python-client) doubled this year. There were
a lot of performance improvements around the integration with the sqlalchemy

> integration. This also included a range of bug fixes.

Reviewer: ```suggestion
integration. This also included a range of bug fixes, and benefits many Python-based tools in the Trino ecosystem.
```

### Airflow integration

The long-awaited [Trino/Airflow integration
](https://airflow.apache.org/docs/apache-airflow-providers-trino/stable/index.html)
landed this year. This paired well with the new task-retry and fault-tolerant

> execution features. There was a small tutorial on [how to use Trino and Airflow

Reviewer: We should probably not link to this links since it does NOT use the proper support that is now available.

Author: Ah, fair

> ]({% post_url 2022-07-13-how-to-use-airflow-to-schedule-trino-jobs %})

Reviewer: wrapping might break link

from Willie Valdez that details how to get started. To learn more about the full
capabilities of pairing Trino’s few fault-tolerant execution mode with Airflow,
check out [Philippe Gagnon’s talk at this year’s Trino Summit
](https://www.youtube.com/watch?v=xKDN7RUJ5i4).

### Metabase driver

> A lot of folks in the community were asking for a [Trino/Metabase driver

Reviewer: wrapping might break link

](https://github.com/metabase/metabase/issues/17532) after Trino updated its

> name. This was a large blocker for anyone who wanted to move over to Trino but

Reviewer: This was a large blocker for anyone who wants to move to Trino and uses Metabase.

> Metabase was the primary BI tool used. Through some collaboration with the

Reviewer: ```suggestion
Metabase was the primary BI tool used. Through a collaboration of the
```

> Metabase and Starburst folks, the [metabase-drive

Reviewer: driver

> ](https://github.com/starburstdata/metabase-driver) for Trino/Starburst made its

Reviewer: for Trino was released, and we saw numerous users migrate to Trino.

way in earlier this year, and we saw quite a few folks migrate over from Presto
with this new open source driver.

## 2023 Roadmap

The upcoming roadmap was [covered in detail](https://youtu.be/mUq_h3oArp4?t=799)
by Martin at Trino Summit. To avoid too extending this blog even further, we’ll
leave you with the featured project that covers many aspects of the Trino core
engine.

### Project Hummingbird

[Project Hummingbird](https://github.com/trinodb/trino/issues/14237) aims to

> improve Trino's columnar and vectorized evaluation engine. Every year we report

Reviewer: storage and evaluation engine 

??

Author: I got this from https://github.com/trinodb/trino/issues/14237 and it seems pretty central to a lot of the efforts behind hummingbird.

on many incremental performance improvements. These improvements are typically

> small in isolation but have a large aggregate impact as they start to compound

Reviewer: drop "as they start ..."

and overlap with more common workloads being run in production. This incremental

> approach is the real key to improving query engine performance, but there is

Reviewer: approach is the real key to improving query engine performance, and there is alway room for further optimization.

> still room for further optimization. It’s time to take Trino to the next level

Reviewer: Could we drop this sentence "It's time ..."

by bringing improvements in areas such as filter, projection, aggregation and
join evaluation, as well as any other potential improvements in areas. If you
want to get involved with this exciting project, or to learn about the latest

> innovations as they are being discussed, joint the #project-hummingbird in

Reviewer: #project-hummingbird channel

[the Trino Slack workspace]({{site.url}}/slack.html).

## Conclusion

> 2022 was by far the busiest this bunny has been. While we’re all kind of waiting

Reviewer: Drop the whole sentence "While we ..." or at least shorten

for the moment where the trajectory of things going on would start to taper off,
it’s only growing as we attract more contributors. We believe that this trend
will continue in 2023 as we begin to put more process in place around managing

> pull requests. We are also in the process of building out a library of Trino

Reviewer: Lets not talk about the user group, presentation and other stuff just yet .. lets announce when we have more in a separate blog post and create a call out for calendar entries, speakers and more

assets, presentations, and other materials that make building out a compelling
presentation or writing a blog post about Trino much easier. Remember to get
the word out and share anything you genuinely think is cool or important for

> others to hear! Let’s get to the point that people talk about Trino, Snowflake,

Reviewer: What do you mean here? Need to clarify and reword
